File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include "math.h"
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        long  a = 0, b = 0;
        long  carry = 0;
        
        if (!l1) return l2;
        if (!l2) return l1;
        ListNode* prev = NULL;
        ListNode* root = NULL;
        while( l1 || l2 || carry)
        {
            int val = 0;
            if (l1 && l2)
                val = (carry) ? l1->val + l2->val + carry : l1->val + l2->val;

            else if (l1 && !l2)
                val = (carry) ? l1->val + carry : l1->val;
            
            else if (l2 && !l1)
                val = (carry) ? l2->val + carry : l2->val;
            
            else
                val = carry;
            
            carry = (val < 10) ? 0 : 1;
            
            if (val >= 10)
            {
                val = val % 10;
            }
            ListNode* node = new ListNode();
            node->val = val;

            if (!prev)
            {
                prev = node;
                root = prev;
            }
            else
            {
                prev->next = node;
                prev = node;
            }
            if (l1)
                l1 = l1->next;
            if (l2)
                l2 = l2->next;
        }
        return root;
    }
};

File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }
        
        int carry = 0;
        ListNode res = null, prev = null;
        while(true) {
            int sum = 0;
            if (l1 == null && l2 == null) {
                sum = carry;
            }
            else if (l1 == null) {
                sum = l2.val + carry;
            }
            else if (l2 == null) {
                sum = l1.val + carry;
            }
            else {
                sum = l1.val + l2.val + carry;                
            }
            
            if (sum >= 10) {
                carry = 1;
                sum = sum % 10;
            }
            else {
                carry = 0;
            }
            
            ListNode node = new ListNode(sum, null);
            
            if (res == null) {
                res = node;
                prev = node;
            }
            else {
                prev.next = node;
                prev = node;
            }
            l1 = l1 == null ? l1 : l1.next;
            l2 = l2 == null ? l2 : l2.next;
            
            if (l1 == null && l2 == null && carry == 0) {
                break;
            }
        }
        return res;
    }
}